[export] Add serialization support for tokens

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121552

Differential Revision: [D54906766](https://our.internmc.facebook.com/intern/diff/D54906766)

angelayi committed:
[export] Add serialization support for tokens

[ghstack-poisoned]

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -2311,6 +2358,8 @@ def rank_input(inp) -> Tuple[int, Optional[str], int]:
         elif spec.type == "custom_obj":
             return 4, spec.custom_obj.custom_obj_name, idx
+        elif spec.type == "token":
+            return 0, None, idx
```
@angelayi is token always ahead of other nodes for inputs and outputs?

angelayi replied:
yeah

zhxchen17 commented on torch/_export/serde/schema.py:
```diff
@@ -236,13 +236,19 @@ class InputToCustomObjSpec:
 
+@dataclass
+class InputTokenSpec:
+    arg: TensorArgument
```
why tokens are tensors instead of their own types?

angelayi replied:
Mm... the tokens themselves are actually just tensors

suo replied:
if they are tensors, do they have associated tensormetas, entries in the tensorvalues table, etc?

angelayi replied:
They have associated tensor meta, but no entries in the tensorValues table since it doesn't matter what value they have -- they can just be dummy `tensor([])`

suo replied:
I find this sort of weird…is there a specific reason why we want to make tokens into tensors instead of their own type?

angelayi replied:
I think it was just easier to introduce them as tensors since tensors are already integrated into everything. I guess we can make it a special type -- is it easy to introduce a new type of input?

suo replied:
It looks like:
1. Need to add a new IValue type (ivalue.h)
2. Need to add schema support, (schema_type_parser.cpp)
3. Need to add it to the export schema
5. Backends now need to implement it.

But I think it's a good idea. Re-using a tensor for this is hacky; it makes the operator schema misleading and creates a new special case that code that operates generically over tensors in the graph needs to be aware of.

@zou3519 thoughts?

angelayi replied:
I updated the existing code to use a new TokenArgument rather than TensorArgument.
And as for adding the new IValue, we can discuss with Richard once hes back from his pto

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -995,32 +1012,50 @@ def serialize_hoo_outputs(self, node: torch.fx.Node) -> List[Argument]:
-
             if len(meta_val) == 1:
                 tensors = []
                 for i, v in enumerate(meta_val):
```
If len(meta_val) == 1, we don't need to iterate on it?

angelayi replied:
It could be a Tensor list

angelayi committed:
Update on "[export] Add serialization support for tokens"




[ghstack-poisoned]

angelayi committed:
Update on "[export] Add serialization support for tokens"




[ghstack-poisoned]

angelayi committed:
Update on "[export] Add serialization support for tokens"


Differential Revision: [D54906766](https://our.internmc.facebook.com/intern/diff/D54906766)

[ghstack-poisoned]

angelayi committed:
Update on "[export] Add serialization support for tokens"


Differential Revision: [D54906766](https://our.internmc.facebook.com/intern/diff/D54906766)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54906766).

angelayi commented:
@pytorchbot merge -f "landed internally"
